server/apps/packages/package_service.py

```diff
@@ -222,6 +229,50 @@ def get_packages(self, doc_id):
 
         return get_resource_service(ARCHIVE).get_from_mongo(req=request, lookup=query)
 
+    def remove_ref_from_inmem_package(self, package, ref_id):
+        """
+        Removes the reference with ref_id from non-root groups. If there is nothing left
+        in that group then the group and its reference in root group is also removed.
+        If the removed item was the last item then returns
+        :param package: Package
+        :param ref_id: Id of the reference to be removed
+        :return: True if there are still references in the package, False otherwise
+        """
+        groups_to_be_removed = []
+        non_root_groups = [group for group in package.get(GROUPS, []) if group.get(GROUP_ID) != ROOT_GROUP]
+        for non_rg in non_root_groups:
+            refs = [r for r in non_rg.get(ASSOCIATIONS, []) if r.get(ITEM_REF, '') != ref_id]
+            if len(refs) == 0:
+                groups_to_be_removed.append(non_rg.get(GROUP_ID))
+            non_rg[ASSOCIATIONS] = refs
+
+        if len(groups_to_be_removed) > 0:
+            root_group = [group for group in package.get(GROUPS, []) if group.get(GROUP_ID) == ROOT_GROUP][0]
+            refs = [r for r in root_group.get(ASSOCIATIONS, []) if r.get(ID_REF) not in groups_to_be_removed]
+            root_group[ASSOCIATIONS] = refs
+            removed_groups = [group for group in package.get(GROUPS, [])
+                              if group.get(GROUP_ID) not in groups_to_be_removed]
+            package[GROUPS] = removed_groups
+
+            # return if the package has any items left in it
+            return len(refs) > 0
+
+        # still has items in the package
+        return True
+
+    def replace_ref_in_package(self, package, old_ref_id, new_ref_id):
+        """
+        Locates the reference with the old_ref_id and replaces with the new_ref_id
+        :param package: Package
+        :param old_ref_id: Old reference id
+        :param new_ref_id: New reference id
+        """
+        non_root_groups = [group for group in package.get(GROUPS, []) if group.get(GROUP_ID) != ROOT_GROUP]
```

Comment:
Hm, what do you mean? Doesn't the next line simply iterate over the `non_root_groups` iterable once?

General answer: yes, it is possible to say `x in generator`, but only if you do it once, because the check exhausts the generator.

```py
>>> g = (a for a in range(5))
>>> g
<generator object <genexpr> at 0x7f909d158500>
>>> 3 in g
True
>>> 3 in g
False
>>> 3 in g
False
```